### ISSUE TITLE
Improve code documentation and variable naming

### ISSUE THREAD

**Issue by LowellMontgomery** [mcrosson/lr_plugin_computer_vision_tagging#8]
This is currently a bit of a blocker for me as I work on my feature branch for hierarchical keywords. Since a name may appear in more than one keyword, we need to add an index (default: 1) to the ID for each checkbox... that way we can have two checkboxes with the same name.

My current issue is looking at code like:

```lua
      for photo, apiResult in pairs(photosToTag) do
        tagsByPhoto[photo] = {};
        local tagDetails = ClarifaiAPI.processTagsProbabilities(apiResult);
        for _, tag in ipairs(tagDetails) do
          tagsByPhoto[photo][tag.tag] = tag;
        end
      end
``` 

With no explanation of what is going on, code like `tagsByPhoto[photo][tag.tag] = tag;` just doesn't make sense. Sure, I can start doing lots of debug output to figure it out, and get an idea where to change the structure of which variables and where to add the extra loops to deal with the new index (since each tag name could correspond to 2 or more checkboxes, not necessarily just one)... but I'm getting confused and I think that indicates an issue that could be better looked at and resolved. The code would be a bit more maintainable (and should be, if you want to attract collaboration) if it were better documented and if variable naming were a bit more clear.

More code that makes my head hurt: ;-) 

```lua
    for tag, taginfo in pairs(tags) do
      if taggingProgress:isCanceled() then
        break;
      end
``` 
[...]

```lua
          local keyword = catalog:createKeyword(taginfo.tag, {}, false, nil, true);
          if keyword == false then -- This keyword was created in the current withWriteAccessDo block, so we can't get by using `returnExisting`.
            keyword = newKeywords[taginfo.tag];
          else
            newKeywords[taginfo.tag] = keyword;
          end
          
          if tagSelectionsByPhoto[photo][taginfo.tag] then
            photo:addKeyword(keyword);
          else
            photo:removeKeyword(keyword);
          end
```

**Comment by mcrosson** [comment]
The first block should probably be the following

```lua
    -- Post processing of API results so we can find tags for a photo and associated details (tag name, probability, etc)
    -- tagsByPhoto is setup as a nested dict that is photo -> tag -> tag details from different APIs   
    for photo, apiResult in pairs(photosToTag) do
        tagsByPhoto[photo] = {};
        local details= ClarifaiAPI.processTagsProbabilities(apiResult); -- Process the Clarifai API results into a format that can be re-used in the tagging dialog a little cleaner
        for _, detail in ipairs(details) do
          tagsByPhoto[photo][detail.tag] = detail;
        end
      end
```

The below code is an inline check to make sure that the plugin bails out if the user cancels the tagging process. Without this check the plugin won't honor a cancel operation.

```lua
    for tag, taginfo in pairs(tags) do
      if taggingProgress:isCanceled() then
        break;
      end
```

This is some magic mojo inspired by SAFX and the fact that keywords created in the current context don't get returned by the catalog:createKeyword call. This shouldn't look too different than what was in SAFX. However the taginfo.tag convention should be "tagDetails" or similar since I use a dict to represent a tag name + additional detail like API probability, API, etc.

```lua
    local keyword = catalog:createKeyword(taginfo.tag, {}, false, nil, true);
          if keyword == false then -- This keyword was created in the current withWriteAccessDo block, so we can't get by using `returnExisting`.
            keyword = newKeywords[taginfo.tag];
          else
            newKeywords[taginfo.tag] = keyword;
          end
          
          if tagSelectionsByPhoto[photo][taginfo.tag] then
            photo:addKeyword(keyword);
          else
            photo:removeKeyword(keyword);
          end
```

**Comment by mcrosson** [comment]
I'll try to improve docs / comments / naming conventions as I work through the code. If you're up for it, let's keep this issue open and I'll try to get explanations as you find confusing code and I'll work on updating the code base as items are brought to my attention.

**Comment by LowellMontgomery** [comment]
What I found confusing about that was not the "progress" calls (been there and done that, at least somewhat, for another plugin), nor the logic for whether/not to create a new keyword. That was clear. But what is taginfo.tag... These are both the index and value of the "tags" table we are iterating over, but how is that rendered (is it a string with a dot in the middle?) Or does "taginfo" have a ".tag" variable? It just starts to look weird and confusing when I look back and forth between the logic changes I had to make in the safx plugin and try to figure out where to make them here. Right now, I have things "working" (but the feature is not added)... it's a part-way point. Perhaps I can give you a pull request so my current "non-broken" state can be in master and changes from you don't end up causing conflicts, then perhaps we could collaborate for how to best make the real functionality changes.

**Comment by LowellMontgomery** [comment]
The only thing I should fix is removing that extra preference for minimum probability for auto-selection, since you added that in the export dialog... then I think that what I have is at a suitable state where we could collaborate based on the beginning changes I've added to provide the discussed functionality.

**Comment by mcrosson** [comment]
>But what is taginfo.tag

tagInfo is a dictionary that contains tag name, api, probability and other information culled from upstream services. I should have refactored the variable to be "apiTagDetail" or similar now that I look at it with fresh eyes. A definition of what the tagInfo dictionary currently looks like:
`{ tag = 'foo', probability = '0.0001', service = 'Clarifai'}`

**Comment by LowellMontgomery** [comment]
Okay, so taginfo.tag is "foo", or the name of the suggested keyword. And elsewhere, where we have tag.tag, that's the same, just a different variable name? (I need to look back at the relevant code blocks). Anyway, thanks for clarifying (no pun intended or I would have spelled it Clarifai-ing... ;-) )

**Comment by mcrosson** [comment]
Correct, that's what the standard "convention" is. Part of the problem is that I never tweaked some of the base variable names as I adjusted the functionality and moved it in a direction that will help support multiple computer vision APIs / services.

As I work through the code in the future I'll try to adjust the variables so it's more obvious what's going on in the code.

**Comment by LowellMontgomery** [comment]
👍  Good.

By the way (somewhat relevant to this ticket), maybe we could add http://www.johnrellis.com/lightroom/debugging-toolkit.htm implementation to this project? If you change the plugin extension name to lrplugin, it ignores the debug stuff entirely, so it's nice for having a way to just output stuff in a clean way. And it's error debugging window is truly gold... If your code gets broken, it's much nicer than what you see without it...

**Comment by LowellMontgomery** [comment]
I used it when I was sorting out some issues in my https://github.com/LowellMontgomery/LrIptcCodeHelper, so you can see examples of it in use. The way I've got it configured, my "Git Checkouts" directory is where both the module and my plugin's project directory are located, but the .lrdevplugin directory is an extra directory deeper (as in your project), since the documentation is not inside the actual plugin. So configuration looks like:

```lua
local Require = require 'Require'.path ("../../debugscript.lrdevplugin")
local Debug = require 'Debug'.init ()
require 'strict'
```

:point_up: That is added to any file that needs debugging and a copy of his Require.lua needs to be added to the project.

**Comment by mcrosson** [comment]
@LowellMontgomery The note about the debugging was a great idea and I've moved the relavant information to it's own support ticket. I'll be adding that to the implementation list "soon".
